contribute/architecture/services.md

```diff
@@ -0,0 +1,183 @@
+
+All services need to implement the [Service](https://godoc.org/github.com/grafana/grafana/pkg/registry#Service) interface:
+
+```go
+type MyService struct {
+}
+
+func (s *MyService) Init() error {
+    return nil
+}
+```
+
+The `Init` method is used to initialize and configure the service to make it ready to use.
+
+### Register a service
+
+Every service needs to be registered for the application to find it.
+
+To register a service, call the `registry.RegisterService` function in a `init` function within your package.
+
+```go
+func init() {
+    registry.RegisterService(&MyService{})
+}
+```
+
+`init` functions are only run whenever a package is imported, so we also need to import the package in the application. In the `server.go` file under `pkg/cmd/grafana-server`, import the package we just created:
+
+```go
+import _ "github.com/grafana/grafana/pkg/services/mysvc"
+```
+
+## Communication
+
+So far, our service is not doing much. Let's change that by having it interact with other services.
+
+Grafana uses a _service bus_ to pass messages from one service to another. The bus helps decouple the services from each other. All communication over the bus happens synchronously.
+
+For our service to access the bus, we need to add it to the `MyService` struct:
+
+```go
+type MyService struct {
+    bus bus.Bus `inject:""`
+}
+```
+
+> Did you notice the struct tag, `inject:""`? Grafana uses the [inject](https://github.com/facebookgo/inject) package to inject implementations during runtime. 
+
+Now we can start sending and receiving messages! In Grafana, services communicate with each other through three types of messages: _events_, _commands_, and _queries_.
+
+### Events
+
+An event is something that happened in the past. Since an event has already happened, you can't change it. Instead, you can react to events by triggering additional application logic to be run, whenever they occur.
+
+> Because they happened in the past, event names are written in past tense, such as `UserCreated`, and `OrgUpdated`.
+
+#### Subscribe to an event
+
+In order to react to an event, you first need to _subscribe_ to it. 
+
+To subscribe to an event, register an _event listener_ in the service's `Init` method:
+
+```go
+func (s *MyService) Init() error {
+    s.bus.AddEventListener(s.UserCreated)
+    return nil
+}
+
+func (s *MyService) UserCreated(event *events.UserCreated) error {
+    // ...
+}
+```
+
+**Tip:** Browse the available events in the `events` package.
+
+#### Publish an event
+
+If you want to let other services react to changes in your service, you can publish your own events:
+
+```go
+event := &events.StickersSentEvent {
+    UserID: "taylor",
+    Count:   1,
+}
+if err := s.bus.Publish(event); err != nil {
+    return err
+}
+```
+
+### Commands
+
+A command is a request for an change to be made. Unlike an event, a command can be declined by the handler for that command. The handler will then return an error.
```

Comment:
I love the fact that we're properly defining what a command is - however, not sure this is what the definition should look like. Commands are not always _changes_ in the system. The way I see it they are just an action to be handled.

On another note (potential dumb question),
>a command can be declined by the handler for that command

I'm not sure what this means in this context. How can the handler decline the execution of a command? Is the case where a handler is not present for the command?

Reply:
Would "request for an action to be taken" work?

I'm trying to convey that unlike events, command are not facts. It's up to the handler to decide whether the command is valid, and whether and any action is taken/change is made. I've seen implementations (outside the Grafana project) where commands are handled as events, which will result in inconsistencies if the handler wasn't able to complete the command.

By decline, I'm referring to handler logic that might decide that the system isn't able to complete the command. For example, if a `DeleteDashboardCommand` is dispatched, but no dashboard with that ID is available, the handler can't fulfill the request. 

Is there a better way to phrase this?

Reply:
> Would "request for an action to be taken" work?

Love it, let's run with that ❤️ 

> Is there a better way to phrase this?

That makes sense! How do you feel about: 

`Unlike an event's fire and forget approach, a command can fail as by its handler.`

Declination of a handler would imply that the command was never received by the handler as opposed to a failure to complete execution.